src/frontend/src/optimizer/rule/index_selection.rs

```diff
@@ -0,0 +1,300 @@
+/// for a in (xxx) and b = 1 and c = 1, its cost is In0 * Equal1 * Equal2 = 10
+/// for a = 1 and b in (xxx), its cost is Equal0 * In1 * All2 = 1 * 8 * 50 = 400
+/// for a between xxx and yyy, its cost is Range0 = 500
+/// for a = 1 and b between xxx and yyy, its cost is Equal0 * Range1 = 50
+/// for a = 1, its cost is 100 = Equal0 * All1 = 100
+/// for no condition, its cost is All0 = 10000
+///
+/// With the assumption that the most effective part of a index is its prefix,
+/// cost decreases as `column_idx` increasing.
+/// for index order key length > 5, we just ignore the rest.
+
+const INDEX_MAX_LEN: usize = 5;
+const INDEX_COST_MATRIX: [[usize; INDEX_MAX_LEN]; 4] = [
+    [1, 1, 1, 1, 1],
+    [10, 8, 5, 5, 5],
+    [500, 50, 20, 10, 10],
+    [10000, 100, 50, 20, 20],
+];
+
+pub struct IndexSelectionRule {}
+impl Rule for IndexSelectionRule {
+    fn apply(&self, plan: PlanRef) -> Option<PlanRef> {
+        let logical_scan: &LogicalScan = plan.as_logical_scan()?;
+        let indexes = logical_scan.indexes();
+        if indexes.is_empty() {
+            return None;
+        }
+
+        let mut primary_table_scan_io_estimator = TableScanIoEstimator::new(logical_scan);
+        let primary_cost = primary_table_scan_io_estimator.estimate(logical_scan.predicate());
+
+        let mut final_plan = logical_scan.clone();
+        let mut min_cost = primary_cost.clone();
+
+        let required_col_idx = logical_scan.required_col_idx();
+        for index in indexes {
+            let p2s_mapping = index.primary_to_secondary_mapping();
+            if required_col_idx.iter().all(|x| p2s_mapping.contains_key(x)) {
+                // covering index selection
+                let index_scan = logical_scan.to_index_scan(
+                    &index.name,
+                    index.index_table.table_desc().into(),
+                    &p2s_mapping,
+                );
+
+                let mut index_table_scan_io_estimator = TableScanIoEstimator::new(&index_scan);
+                let index_cost = index_table_scan_io_estimator.estimate(index_scan.predicate());
+                if index_cost.le(&min_cost) {
+                    min_cost = index_cost;
+                    final_plan = index_scan;
+                }
+            } else {
+                // TODO: non-covering index selection
+            }
+        }
+
+        if min_cost == primary_cost {
+            None
+        } else {
+            Some(final_plan.into())
+        }
+    }
+}
+
+struct TableScanIoEstimator<'a> {
+    table_scan: &'a LogicalScan,
+    row_size: usize,
+}
+
+impl<'a> TableScanIoEstimator<'a> {
+    pub fn new(table_scan: &'a LogicalScan) -> Self {
+        Self {
+            table_scan,
+            row_size: table_scan
+                .table_desc()
+                .columns
+                .iter()
+                .map(|x| TableScanIoEstimator::estimate_data_type_size(&x.data_type))
```

Comment:
should we estimate the pk columns 2 times? because the pk column appears in both key and values in our current encoding. cc @fuyufjh

Reply:
Actually the idea comes from weight function of the greedy heuristics join reorder algorithm. Weight function will give a high weight for condition like `Equal` to other condition or no condition for join reorder which is similar to
 our case. Theoretically, there are infinite number of weights can achieve our goal. The pros is we can tune it the weight to affect the algorithm if some unexpected corner case occurs later. The cons is a little bit tricky.

Reply:
I will estimate the pk columns twice if we need in the later PR. I think either one or twice make no huge difference. Especially if we implement the dedup pk encoding in the future.